README.md
# nude crawler

Nude crawler crawls all pages on telegra.ph for today and N past days for specific words, checks number of nude, non-nude images, videos (not analysed) and reports pages which looks interesting (e.g. has more then 10 nude images, or has one video)

## Ineffective intriguing warning
No matter how old you are, no matter how tolerant you are, no matter what your sexual orientation is, no matter what your favorite perversion is, no matter how big your sexual horizons are, with NudeCrawler you will find a lot of things that **you will NOT like**.

I wrote this warning because I have seen some shit. LITERALLY.

Please use it only for legal and ethical purposes. And it's 18+ surely. 

## Install

Recommended (and most secure) way is using docker:
```
mkdir /tmp/run
sudo docker run --rm -v /tmp/run:/work yaroslaff/nudecrawler nudecrawler -a Eva "Sasha Grey" "Belle Delphine" Amouranth
```

See below how to refine your searching and filtering.

### Alternative install
```
pip3 install nudecrawler
```



or, install right from git repo:
```
pip3 install git+https://github.com/yaroslaff/nudecrawler
```

## Launch Nude Crawler!

(I intentionally changed links, do not want to violate github policy)
~~~
$ nudecrawler sasha-grey
INTERESTING https://telegra.ph/sasha-grey-XXXXXXXX
  Nude: 0 non-nude: 0
  Total video: 1

INTERESTING https://telegra.ph/sasha-grey-XXXXX
  Nude: 9 non-nude: 6

INTERESTING https://telegra.ph/sasha-grey-XXXXX
  Nude: 9 non-nude: 6

INTERESTING https://telegra.ph/sasha-grey-XXXXX
  Nude: 6 non-nude: 3
~~~

## Refine search/filtering
Nudecrawler uses [evalidate](https://github.com/yaroslaff/evalidate) to filter results with python expression (`--expr`). With `-h` help will list all avaliable variables, like: `total_images nude_images nonnude_images new_nude_images new_nonnude_images new_total_images total_video`.
Default value: `(total_images>5 and new_nude_images>0) or total_video>0`.

Use `-a`/`--all` to get some results ASAP (but later you may want to make some filtering)

Consider using `--days`, `--total` to narrow/wider search.

Also, `--cache`, `--max-pictures`, `--max-errors`, `--min-content-length` and `--minsize` to speed-up searching and discard some images/pages before wasting time on it.


### Long-time run

#### Stop/Resume
When working with worklists an --stats file, current status is periodically saved to this file. If you need to resume it, just use command `nudecrawler --resume PATH/TO/stats.json`

#### Memory leaking in containers
You may check container memory usage with `sudo docker stats` or `sudo docker stats --no-stream`. Often containers consume more and more memory with time, leading to out-of-memory in the end. To prevent this problem use combination of `--stop` and `--refresh` like `--stop 1000 --refresh bin/refresh-nsfw-api.sh` this will call refresh script every 1000 images. Refresh script should stop current container and start it again. See source of refresh-nsfw-api.sh for example, it's very simple.

### Benchmarking/test
Tested on same page, different technologies (default thresholds) gives different results.
Page A: *belle delphine from 16th Jan* (64 lite sexy images, mostly underwear, nude breast on few)
Page B: *sasha grey* from 18 Apr (16 images, 12 clearly nsfw, 4 are clearly safe )

| filtering technology           | A time | A nudes | B time | B nudes                            |
|---                             |---     | --      |---     |---                                 | 
|:nude (bilt-in)                 | 127s   | 63      | 34s    | 14 (false positives/negatives)     |
|detect-image-nsfw_api (docker)  | 90s    | 49      | 23s    | 12                                 |
|detect-image-aid (docker)       | 124s   | 10      | 28s    | 6 (false negatives)                |
|detect-image-nudenet  (scripts) | 90s    | 57      | 24s    | 12                                 |

## Options
~~~
usage: nudecrawler [-h] [-d DAYS] [--url1 URL] [-f FAILS] [--day MONTH DAY] [--expr EXPR] [--total N] [--max-errors N] [--min-content-length N] [-a] [--detect-image SCRIPT]
                   [--detect-url SCRIPT] [--detect METHOD] [--extensions [EXTENSIONS ...]] [--minsize MINSIZE] [--max-pictures N] [--cache PATH] [-v] [--unbuffered] [--urls] [--log LOG]
                   [--bugreport] [--workdir WORKDIR] [-w WORDLIST] [--stats STATS_FILE] [--resume STATS_FILE] [--stop NUM_IMAGES] [--refresh SCRIPT [ARG ...]]
                   [words ...]

Nudecrawler: Telegra.ph Spider 0.3.6
https://github.com/yaroslaff/nudecrawler

positional arguments:
  words

optional arguments:
  -h, --help            show this help message and exit
  -d DAYS, --days DAYS
  --url1 URL            process only one url
  -f FAILS, --fails FAILS
                        stop searching next pages with same words after N failures
  --day MONTH DAY       Current date (default is today) example: --day 12 31
  --expr EXPR, -e EXPR  Interesting if EXPR is True. def: '(total_images>5 and new_nude_images>0) or total_video>0'
                        Fields: total_images nude_images nonnude_images new_nude_images new_nonnude_images new_total_images total_video
  --total N             Boring if less then N total images (5)
  --max-errors N        Max allowed errors on page ()
  --min-content-length N
                        Interesting if N+ total images (5)

Image filtering options:
  -a, --all             do not detect, print all found pages
  --detect-image SCRIPT
                        explicitly use this script to detect nudity on image file
  --detect-url SCRIPT   explicitly use this script to detect nudity on image URL
  --detect METHOD       One of true, false, nudepy, nudenetb, aid, nsfwapi, nudenet
  --extensions [EXTENSIONS ...]
                        interesting extensions (with dot, like .jpg)
  --minsize MINSIZE     min size of image in Kb (10)
  --max-pictures N      Detect only among first prefiltered N pictures
  --cache PATH          path to cache file (will create if missing)

Output options:
  -v, --verbose         verbose
  --unbuffered, -b      Use unbuffered stdout
  --urls                Do not detect, just generate and print URLs
  --log LOG             print all precious treasures to this logfile
  --bugreport           send bugreport in case of problem (works only after agreed in github ticket)
  --workdir WORKDIR     Use all files (log, wordlist, cache) in this dir. def: .

list-related options:
  -w WORDLIST, --wordlist WORDLIST
                        wordlist (urllist) file
  --stats STATS_FILE    periodical statistics file
  --resume STATS_FILE   resume from STATS_FILE (other args are not needed)
  --stop NUM_IMAGES     stop (or --refresh) after N images processed (or little after)
  --refresh SCRIPT [ARG ...]
                        run this refresh script every --stop NUM_IMAGES images
~~~

## Advanced Usage

### Working with wordlists
In simplest case (not so big wordlist), just use `-w`, like:
~~~shell
# verbose, no-filtering (report all pages), use wordlist
nudecrawler -v -a -w wordlist.txt
~~~

If you have very large wordlist, better to pre-check it with faster tool like [bulk-http-check](https://github.com/yaroslaff/bulk-http-check), it's much faster, doing simple check (we need only filter-out 200 vs 404 pages) millions of page per hour on smallest VPS server.

Convert wordlist to urllist
~~~shell
# only generate URLs 
nudecrawler -v -w wordlist.txt --urls > urls.txt
~~~
Verify it with [bulk-http-check](https://github.com/yaroslaff/bulk-http-check) and get output file with this format:
~~~
https://telegra.ph/abazhurah-02-26 OK 404
https://telegra.ph/ab-03-01 OK 200
https://telegra.ph/aaronov-02-22 OK 404
https://telegra.ph/abazhurami-02-25 OK 404
~~~

Filter it, to leave only existing pages, and strip date from it:
~~~
grep "OK 200" .local/urls-status.log | cut -f 1 -d" "| sed 's/-[0-9]\+-[0-9]\+$//g' | sort | uniq > .local/urs.txt
~~~

List (urls.txt) will look like:
~~~
https://telegra.ph/
https://telegra.ph/a
https://telegra.ph/ab
https://telegra.ph/aba
https://telegra.ph/Abakan
....
~~~
This list (~300Kb, 11k urls) created from 1.5M words russian wordlist. There are only words which had at least one page with this title for last 10 days. So it has words 'Анжелика' or 'Анфиса' (beautiful woman names), but has no words 'Абажурами' or 'Абажуродержателем' (Because there are no pages with these titles on telegra.ph).

Now you can use this file as wordlist (nudecrawler will detect it's already base URL, and will only append date to URL). 


### JSON log files
If `--log` filename ends with `.json` or `.jsonl`, nudecrawler will save log in JSONL format (each line is JSON for a page). Example:

~~~
{"status": "INTERESTING", "url": "https://telegra.ph/Masha-NN-NN-N", "total_images": 29, "nude_images": 17, "new_nude_images": 17, "nonnude_images": 12, "new_nonnude_images": 12, "total_video": 0}
{"status": "INTERESTING", "url": "https://telegra.ph/Masha-NN-NN-N", "total_images": 3, "nude_images": 3, "new_nude_images": 3, "nonnude_images": 0, "new_nonnude_images": 0, "total_video": 0}
~~~

You can save almost all pages and then filter it with jq (get only interesting records, only interesting fields):
~~~
jq 'select(.nude_images>1 and .total_images>1) | {"url": .url, "total": .total_images}' < /tmp/n.json
~~~


### Working with different nudity detectors

NudeCrawler can work with different nudity detectors and very easy to extend. Option `-a`/`--all` will disable detection totally, and it will report all pages.

Bult-in filter `:nude` based on [nude.py](https://github.com/hhatto/nude.py), (python port of [nude.js](https://github.com/pa7/nude.js)) is mostly good and used by default (and does not needs to install many dependecties as with keras/tensorflow detectors, which better to use as Docker images), but it's slower

There are two options to connect user filters, `--detect-image SCRIPT` and `--detect-url SCRIPT`, first one will call script and pass it filename of downloaded image to analyse, and second one will call script and pass it URL of image to analyse. Script should return with either 0 return code (image is SFW) or 1 (image is NSFW). Mnemonic: return code is number of *interesting* images. 

if you will use `/bin/true` as script, it will detect all images as nude, and `/bin/false` will detect all images as non-nude.

Scripts are usually installed to /usr/local/bin and if it's in $PATH, you do not need to specify full path to script, nudecrawler will find it in $PATH.

#### detector: nsfw_api (recommended)

To use [nsfw_api](https://github.com/arnidan/nsfw-api):

Start:
~~~
sudo docker run --rm --name nsfw-api -d -p 3000:3000 ghcr.io/arnidan/nsfw-api:latest
~~~

Use option `--detect nsfwapi`

This detector understands DETECTOR_VERBOSE, and special threshold for each of NSFW classes (porn, sexy, hentai),
also, DETECTOR_THRESHOLD sets default threshold for all classes.
~~~
DETECTOR_VERBOSE=1 DETECTOR_THRESHOLD_HENTAI=0.9 bin/detect-image-nsfw-api.py /tmp/sketch-girl.jpg ; echo $?
Safe /tmp/sketch-girl.jpg: {'hentai': 0.57, 'drawing': 0.4, 'porn': 0.02, 'neutral': 0.01, 'sexy': 0.0}
0
~~~

#### detector: adult-image-detector 
To use [adult-image-detector](https://github.com/open-dating/adult-image-detector):
~~~
sudo docker run --rm -d -p 9191:9191 --name aid --memory=1G opendating/adult-image-detector
~~~

And use option `--detect aid`

adult-image-detector works good and fast for me, but has memory leaking so needs more and more RAM. It's good for short-time run

#### detector: NudeNet

##### Installing NudeNet (little trick needed)
Using NudeNet does not requires docker, but you need to install `pip3 install -U flask nudenet python-daemon` (consider using virtualenv, because nudenet has many dependencies). Also, NudeNet requires model in file `~/.NudeNet/classifier_model.onnx`, if file is missing, NudeNet (unsuccessfully) *tries* to download file from https://github.com/notAI-tech/NudeNet/releases/download/v0/classifier_model.onnx but there is problem, github may display warning page instead of real .onnx file, so this page is downloaded (which is certainly wrong).

Right way workaround is simple - after you will install NudeNet download model *manually* (no wget!) and place it to `~/.NudeNet/`

Or you can download from my temporary site: `wget https://nudecrawler.netlify.app/classifier_model.onnx` (But I cannot promise it will be there forever) and put it to ~/.NudeNet .


##### Using NudeNet with NudeCrawler
[NudeNet](https://github.com/notAI-tech/NudeNet) filtering is implemented as client-server. Start server (PATH_TO/detect-server-nudenet.py) on other terminal (or screen/tmux) and add option `--detect nudenet` to NudeCrawler.

#### Writing your own detector
If you want to write your own detector, explore current detector scripts as example, but here is main rules:
- Image URL or PATH passed as argv[1]
- Return 0 if image is safe and boring, return 1 if image is interesting
- Return 0 if there are any technical problems (timeout or 404)
- Additional configuration could be specified via environment, NudeCrawler will pass environment to your script
- NudeCrawler also sets env variables `NUDECRAWLER_PAGE_URL` and `NUDECRAWLER_IMAGE_URL`

### Building docker container
Repository includes Dockerfile. Use `sudo docker build --build-arg VERSION=0.3.10 -t yaroslaff/nudecrawler:0.3.10 -f docker/Dockerfile .` to build it.

Running docker container (example):
~~~
mkdir /tmp/run
sudo docker run -v /tmp/run:/work yaroslaff/nudecrawler nudecrawler -a Eva "Sasha Grey" "Belle Delphine" Amouranth
~~~

If you specify files for docker (like `-w`, `--stats`, `--resume`, `--log`, `--cache`) path will be modified starting from /work. e.g. `-w urls.txt` will be `-w /work/urls.txt` which is /tmp/run/urls.txt on host.


### Little bit about internals

#### Prefiltering
To speed-up processing, nudecrawler uses pre-filtering, HTTP HEAD request is performed for any image, and further processing is performed only if images passes basic check:
- Image URL must return status 200
- If server responds with Content-Length in response headers (telegra.ph uses Content-Length), it must be more then `--minsize` (minsize specified in Kb, and default value is 10Kb). This saves us from downloading/filtering icons.


CHANGELOG.md
# 0.2.4 (2023-04-03)
- Count number of new (not found in cache) total/nude/nonnude images
- init stats keys if loaded from stats file of older version
- urls.txt converted to lowercase (and about duplicates are removed)
- unified verbose (-v) print (vprint)

# 0.3.3 (2023-04-03)
- evalidate support 
- persistent cache
- new built-in fake detectors true and false
- very basic pytest test added
- fixed bug in transliteration RU->EN
- detect-server-nudenet.py can work as daemon now
- autostart(/stop) nudenet server
- dotenv

# 0.3.4
- --bugreport

# 0.3.6
- Dockerfile
- --workdir option
- --max-pictures
- built-in optional nudenetb detector (if nudenet is installed)

# 0.3.7
- NUDE_TOTAL

# 0.3.8
- typo fixed, README

# 0.3.11
- fixed --workdir / --resume in docker image

# 0.3.13
- Catch errors (and not crash) with broken images
- Load nudenet classifier only when needed

# 0.3.14
- suppress logging from nudenetb

# 0.3.16
- conditionally print number of new images

# 0.3.17
- Better cache saving: save cache to tmp file and rename, nice error if json damaged, --cache-save option

# 0.3.19
- JSONL support for output log


nudecrawler/page.py

from bs4 import BeautifulSoup
import urllib.request
from urllib.parse import urljoin
from .remoteimage import RemoteImage
from . import verbose
from .verbose import printv
from .exceptions import *
from .cache import cache

from evalidate import evalidate, EvalException

import requests
import hashlib
import http
import time
import json


from urllib.parse import urlparse
import subprocess
import sys
import os

#def trivial_iterator(xx):
#    for x in xx:
#        yield x

processed_images = 0
context_fields = ['total_images', 'nude_images', 'nonnude_images', 'new_nude_images', 'new_nonnude_images', 'new_total_images', 'total_video']

def get_processed_images():
    return processed_images

def sha1sum(path):
    sum = hashlib.sha1()
    with open(path, 'rb') as source:
        block = source.read(2**16)
        while len(block) != 0:
            sum.update(block)
            block = source.read(2**16)
    return sum.hexdigest()


class Page:

    def __init__(self, url: str, all_found=False, detect_image=None, min_total_images = 0, min_images_size = 10*1024,
                 image_extensions=None, max_errors = None, max_pictures=None,
                 detect_url=None, min_content_length=None, ignore_content_length=None, expr='True'):
        self.url = url
        self.nban_links = 0
        self.nban_images = 0
        self.nude_images = 0
        self.nonnude_images = 0        
        self.total_images = 0

        # images not found in cache
        self.new_nude_images = 0
        self.new_nonnude_images = 0        
        self.new_total_images = 0

        self.total_video = 0
        self.text_found = list()
        self.links = list()        
        self.images = list()
        self.videos = list()
        self.all_found = all_found
        self.ignore_content_length = ignore_content_length
        
        
        # minor errors
        self.error_counter = 0

        self.detect_image = detect_image
        self.detect_url = detect_url
        self.image_extensions = image_extensions or ['.jpg', '.jpeg']
        self.min_image_size = min_images_size
        self.min_total_images = min_total_images
        self.max_errors = max_errors
        self.max_pictures = max_pictures

        # expr to filter interesing
        self._code = None

        self.http_code = None

        self._ignore = False # Ignore this page, we think it's spam, duplicate        
        self._status = None
        self._status_detailed = None
        self._status_logged = False
        self._log = list()
        self.check_time = None
        self.content_length = None

        # can throw evalidate.EvalExpression here
        node = evalidate(expr)
        self._code = compile(node, '<user filter>', 'eval')
        
        printv("Processing:", self.url)

        try:
            page = urllib.request.urlopen(self.url)
            self.http_code = page.getcode()
            self.content_length = page.headers.get('content-length')
            if self.content_length is not None:
                self.content_length = int(self.content_length)

            if self.content_length and min_content_length and self.content_length < min_content_length:
                self.ignore(f"content-length {self.content_length} < minimal {min_content_length}")
                return
        except (urllib.error.URLError, ConnectionError, http.client.RemoteDisconnected) as e:
            if hasattr(e, 'status') and e.status == 404:
                # print(e, type(e))
                # silent ignore most usual error (unless verbose)
                printv(url, 404)
                self._status = "IGNORED"
                self._status_detailed = "404"
                self._ignore = True
                self.http_code = e.status
            else:
                if hasattr(e, 'status'):
                    self.http_code = e.status
                self.ignore(f'Exception {e} with {self.url}')            
            return
        self.content_length = page.headers.get('content-length')
        
        if self.ignore_content_length is not None and self.ignore_content_length == self.content_length:
            self.ignore(f"Ignore because matches prev page content-length = {self.content_length}")
            return
        
        self.soup = BeautifulSoup(page, "html.parser")


    def ignore(self, reason):
        self._ignore = True        
        self._status = f"IGNORED"
        self._status_detailed = reason
        printv(f"IGNORE {self.url}, {self._status_detailed}")
        self.log(self._status_detailed)


    def log(self, msg, really=True):
        if not really:
            return        
        self._log.append(msg)

    def check_all(self):
        started = time.time()

        if self._ignore:            
            return
        
        self.total_images = len(self.soup.findAll('img'))
        self.total_links = len(self.soup.findAll('a'))

        self.check_images()
        self.check_video()
    
        self.check_time = round(time.time() - started, 2)
        self.log(f"Check time: {self.check_time}")
        # set status
        self.status()
        

    def check_video(self):
        self.total_video = len(self.soup.findAll('video'))
        for img in self.soup.findAll('video'):
            src = img.get('src')
            self.videos.append(src)

    def error(self, msg):
        self.error_counter += 1
        self.log(f"minor error ({self.error_counter}): {msg}")
        if self.max_errors is not None and self.error_counter > self.max_errors:
            self.ignore(f'Too many minor errors: {self.max_errors}')            

    def prefilter_image(self, url):
        """ True is we should check, False if we can ignore this image"""
        
        if self._ignore:
            return False
        verdict = cache.url2v(url)
        
        if verdict is not None:
            # self.log(f'{url} passed prefilter because cached')
            return True 
        
        path = urlparse(url).path
        ext = os.path.splitext(path)[1]
        if ext not in self.image_extensions:            
            self.log(f"{url} bad extension, ignore")
            return False
        try:
            r = requests.head(url, timeout=1)
        except requests.exceptions.RequestException as e:            
            self.error(f'{url} request exception: {e}')
            return False

        if r.status_code != 200:
            self.error(f"Bad status code: {url} {r.status_code}")
            return False
        try:
            cl = int(r.headers['Content-Length'])
        except KeyError:
            cl = None

        # self.log(f"{url} status:{r.status_code} content-length: {cl}")
        if cl is not None and cl < self.min_image_size:
            self.log(f"Too small image ({int(r.headers['Content-Length'])})")
            return False

        # self.log(f"{url} image passed prefilter")
        return True


    def do_detect_url(self, url):

        if self.detect_url == ':true':
            self.nude_images += 1
            return True
        
        if self.detect_url == ':false':                
            return False            

        rc = subprocess.run([self.detect_url, url])
        if rc.returncode >= 100:
            print("FATAL ERROR")
            sys.exit(1)
        if rc.returncode:
            self.log(f"{url} is nude")
            self.nude_images += 1
        else:
            self.log(f"{url} is NOT nude")
            self.nonnude_images += 1
        
        return 

    def do_detect_image(self, url):
        try:
            ri = RemoteImage(url)
            sum = sha1sum(ri.path)
            verdict = cache.sum2v(sum, url=url)
            if verdict is not None:                    
                if verdict:
                    self.log(f"{url} is nude (cached sum)")
                    self.nude_images += 1
                else:
                    self.log(f"{url} is NOT nude (cached sum)")
                    self.nonnude_images += 1
                return verdict

            verdict = ri.detect_image(self.detect_image)                
            self.new_total_images += 1

            if verdict:
                self.new_nude_images += 1
            else:
                self.new_nonnude_images += 1
            
            cache.register(url, sum, verdict)
            if verdict:
                self.log(f"{url} is nude")
                self.nude_images += 1
            else:
                self.log(f"{url} is NOT nude")                    
                self.nonnude_images += 1
            return 
        except NudeCrawlerException as e:
            print(e)
            return

    def detect_cache_url(self, url):
        verdict = cache.url2v(url)
        if verdict is not None:
            if verdict:
                self.log(f"{url} is nude (cached url)")
                self.nude_images += 1
            else:
                self.log(f"{url} is NOT nude (cached url)")
                self.nonnude_images += 1
        return verdict
        

    def is_nude(self, url):
        os.environ["NUDECRAWLER_PAGE_URL"] = self.url
        os.environ["NUDECRAWLER_IMAGE_URL"] = url

        if self.detect_cache_url(url) is not None:
            return

        if self.detect_url:
            self.do_detect_url(url)
            return

        if self.detect_image:
            try:
                self.do_detect_image(url)
            except Exception as e:
                printv("Broken image:", url)
            return
                
        
        raise(Exception('Need either detect_image or detect_url or all'))

    def check_images(self):
        global processed_images

        image_list = [ urljoin(self.url, img.get('src')) for img in self.soup.findAll('img') ]
        self.log(f"total images on page: {len(image_list)}")
        
        if len(image_list) < self.min_total_images:
            self.ignore(f'Skip because total images {len(image_list)} < {self.min_total_images}')
            return
        
        image_list = list(filter(None, [ url if self.prefilter_image(url) else None for url in image_list ]))

        if self._ignore:            
            return

        self.log(f"total prefiltered images on page: {len(image_list)}")
        if len(image_list) < self.min_total_images:            
            self.ignore(f'Skip because total prefiltered images {len(image_list)} < {self.min_total_images}')
            return

        if not self.all_found:
            for url in image_list[:self.max_pictures]:
                self.is_nude(url)
                processed_images += 1


    def status(self):        

        if not self._status_logged:
            self._status_logged = True
            makelog = True
        else:
            makelog = False
        
        if self._ignore:
            return f"{self._status} ({self._status_detailed})"

        self._status_detailed = f"total: {self.total_images} (min: {self.min_total_images}) " \
                    f"nude: {self.nude_images}"

        self._status_detailed += f" Cache new/nude: {self.new_total_images}/{self.new_nude_images}"

        if self.total_video:
            self._status_detailed += f" video: {self.total_video}"
        self._status_detailed += f" ({self.check_time}s)"
        
        self.log(self._status_detailed, makelog)

        if self.all_found:
            self._status = "INTERESTING (ALL)"
            self.log(self._status, makelog)        
            return self._status

        # Interesting or not? use evalidate
        ctx = dict()
        for field in context_fields :
            ctx[field] = getattr(self, field)
        r = eval(self._code, ctx.copy())
        if r:
            self._status = f'INTERESTING'
            self.log(self._status, makelog)        
            return self._status            


        self._status = "???"
        self.log(self._status, makelog)        
        return self._status            

    def as_json(self):
        j = dict()
        j['status'] = self.status()

        for attr in ['url', 'total_images', 'nude_images', 'new_nude_images', 'nonnude_images', 'new_nonnude_images', 'total_video']:
            j[attr] = getattr(self, attr)

        return json.dumps(j)

    def __str__(self):
        
        text = ''
        
        text += f'{self.status()} {self.url} ({self.check_time}s)\n'

        if self.all_found:
            text += f"  Total images: {self.total_images}\n"
        else:
            text += f"  Nude: {self.nude_images} ({self.new_nude_images} new) non-nude: {self.nonnude_images} ({self.new_nonnude_images} new)\n"
        
        #if self.new_nude_images or self.new_nonnude_images:
        #    text += f"  New nude {self.new_nude_images} non-nude {self.new_nonnude_images}\n"
                
        if self.total_video:
            text += f"  Total video: {self.total_video}\n"
        
        return text

